clippy_lints/src/lib.rs

```diff
@@ -50,103 +46,23 @@ use rustc_data_structures::fx::FxHashSet;
 use rustc_lint::LintId;
 use rustc_session::Session;
 
-/// Macro used to declare a Clippy lint.
-///
-/// Every lint declaration consists of 4 parts:
-///
-/// 1. The documentation, which is used for the website
-/// 2. The `LINT_NAME`. See [lint naming][lint_naming] on lint naming conventions.
-/// 3. The `lint_level`, which is a mapping from *one* of our lint groups to `Allow`, `Warn` or
-///    `Deny`. The lint level here has nothing to do with what lint groups the lint is a part of.
-/// 4. The `description` that contains a short explanation on what's wrong with code where the
-///    lint is triggered.
-///
-/// Currently the categories `style`, `correctness`, `complexity` and `perf` are enabled by default.
-/// As said in the README.md of this repository, if the lint level mapping changes, please update
-/// README.md.
-///
-/// # Example
-///
-/// ```
-/// #![feature(rustc_private)]
-/// extern crate rustc_session;
-/// use rustc_session::declare_tool_lint;
-/// use clippy_lints::declare_clippy_lint;
-///
-/// declare_clippy_lint! {
-///     /// **What it does:** Checks for ... (describe what the lint matches).
-///     ///
-///     /// **Why is this bad?** Supply the reason for linting the code.
-///     ///
-///     /// **Known problems:** None. (Or describe where it could go wrong.)
-///     ///
-///     /// **Example:**
-///     ///
-///     /// ```rust
-///     /// // Bad
-///     /// Insert a short example of code that triggers the lint
-///     ///
-///     /// // Good
-///     /// Insert a short example of improved code that doesn't trigger the lint
-///     /// ```
-///     pub LINT_NAME,
-///     pedantic,
-///     "description"
-/// }
-/// ```
-/// [lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints
 #[macro_export]
 macro_rules! declare_clippy_lint {
```

Comment:
I'm hesitant to make our `declare_clippy_lint` macro public. If someone uses the clippy_utils outside of Clippy, they should not define a Clippy-style lint.

Reply:
What would be the harm in that?

I am hopeful that this to-be-written tool will maintain lint-level compatibility with Clippy. That way, if a person develops a lint that they think could be of general use, there would be a direct path for inclusion into Clippy. Is this idea flawed?

Reply:
If you develop an extra tool, **it is not Clippy**. So prefixing your lint names with `clippy::` is misleading at best. You can still do this by using `declare_tool_lints`, but we shouldn't provide a convenient way and encourage putting external tool lints under the namespace `clippy::`. 

We already get issues related to rustc lints raised in the Clippy repo, just because they are lints and Clippy is the Rust lint tool. I really don't want to deal with issues coming from external tools, just because they use the `clippy::` namespace, because we provide a convenient way to do so.

Reply:
OK. I see your point. I'll move things around.

Reply:
I think this should be fixed now.